ansible_base/lib/utils/models.py

```diff
@@ -97,3 +99,101 @@ def current_user_or_system_user():
+          this entry is non-empty, it has the form: {"field_name": value} where
+          value is the new value of the field.
+        - removed_fields: A dictionary of fields that were removed between old
+          and new. Similar to added_fields, if old and new are the same type of
+          model, this should always be empty. It has the same structure as
+          added_fields except the value is the old value of the field.
+        - changed_fields: A dictionary of fields that were changed between old
+          and new. The key of each entry is the field name, and the value is a
+          tuple of the old value and the new value.
+    """
+
+    def skip_field(source, field):
+        if field in exclude_fields:
+            return True
+        if limit_fields and field not in limit_fields:
+            return True
+        if not include_m2m and source._meta.get_field(field).many_to_many:
+            return True
+        return False
+
+    diff_dict = {"added_fields": {}, "removed_fields": {}, "changed_fields": {}}
+
+    if old is None and new is None:
+        return diff_dict
+
+    if (old is not None and not isinstance(old, Model)) or (new is not None and not isinstance(new, Model)):
+        raise TypeError('old and new must be a Model instance or None')
+
+    if old is None:
+        for field in get_all_field_names(new, concrete_only=True):
+            if skip_field(new, field):
+                continue
+            diff_dict["added_fields"][field] = make_json_safe(getattr(new, field)) if json_safe else getattr(new, field)
+        return diff_dict
+    elif new is None:
+        for field in get_all_field_names(old, concrete_only=True):
+            if skip_field(old, field):
+                continue
+            diff_dict["removed_fields"][field] = make_json_safe(getattr(old, field)) if json_safe else getattr(old, field)
+        return diff_dict
+    elif require_type_match and type(old) != type(new):  # noqa: E721
+        raise TypeError('old and new must be of the same type')
+
+    for field in get_all_field_names(old, concrete_only=True):
+        if skip_field(old, field):
+            continue
+
+        old_value = getattr(old, field)
+
+        if not hasattr(new, field):
+            diff_dict["removed_fields"][field] = make_json_safe(old_value) if json_safe else old_value
+            continue
+
+        new_value = getattr(new, field)
+        if old_value != new_value:
+            diff_dict["changed_fields"][field] = (
+                make_json_safe(old_value) if json_safe else old_value,
+                make_json_safe(new_value) if json_safe else new_value,
+            )
+
+    for field in get_all_field_names(new, concrete_only=True):
+        if skip_field(new, field):
+            continue
+        if not hasattr(old, field):
+            diff_dict["added_fields"][field] = make_json_safe(getattr(new, field)) if json_safe else getattr(new, field)
+
+    return diff_dict
```

Comment:
What about encrypted fields?

Reply:
I didn't see any reason to handle them differently. Store their changes and if need be someone can decrypt the values in an audit, right?